api.go

```diff
@@ -21,12 +21,20 @@ const timeout = 30 * time.Second
 
 // NewAPI creates a CacophonyAPI instance and obtains a fresh JSON Web
 // Token. If no password is given then the device is registered.
-func NewAPI(serverURL, group, deviceName, password string) (*CacophonyAPI, error) {
+func NewAPI(serverURL, group, deviceName, userName, password string) (*CacophonyAPI, error) {
+	isDevice := deviceName != ""
+	var name string
+	if isDevice {
+		name = deviceName
+	} else {
+		name = userName
+	}
 	api := &CacophonyAPI{
-		serverURL:  serverURL,
-		group:      group,
-		deviceName: deviceName,
-		password:   password,
+		serverURL: serverURL,
+		group:     group,
+		name:      name,
+		isDevice:  isDevice,
```

Comment:
Instead of having having lots of if branches throughout the code based around `isDevice`, what about precomputing the various URLs here and storing those as fields? That makes the rest of the code cleaner and avoids the need for methods like getRegURL.

With this approach, `regURL`, `authURL`, `type` etc would all be calculated and stored in `NewAPI`.

It looks like `isDevice` might still be needed but the number of branches throughout the code will be greatly reduced.